페이먼츠 Step 1 중간 제출합니다.

hyo-choi opened the pull request:
목표한 만큼 못 했네요 ㅠㅠ 되어있는 만큼만 봐주세요! 늘 감사합니다!!

hyo-choi committed:
chore: 프로젝트 환경 설정

hyo-choi committed:
chore: eslint, tsc, husky 설정 추가 및 갱신

- eslint에 lint 룰 추가
- tsconfig 추가
- pre-commit 명령어를 yarn lint로 변경

hyo-choi committed:
feat: Button, Input, Span component 작성

hyo-choi committed:
feat: Card, InputContainer component 작성

Squashed commit of the following:

commit 2c86a19fe9d18c0d46a27f5a300f9f60679eae3a
Author: hyo-choi <hyochoi31@gmail.com>
Date:   Thu Jul 8 00:19:42 2021 +0900

    feat: Card, InputContainer component 작성

commit 9092a39173a732f52511c4fd4c9476f5b66c93dd
Author: hyo-choi <hyochoi31@gmail.com>
Date:   Mon Jul 5 20:14:43 2021 +0900

    feat: Card, InputContainer 작성중

hyo-choi committed:
feat: CardRegisterForm component 작성

hyo-choi committed:
feat: useRef를 이용한 오토포커싱 적용 및 onChange 핸들러 설정

hyo-choi committed:
refactor: CardRegisterForm 컴포넌트 일부 분리, onSubmit 처리

hyo-choi committed:
refactor: input label에 for 추가

hyo-choi committed:
fix: import 경로 수정

hyo-choi committed:
feat: 카드사 선택 모달 추가

hyo-choi committed:
refactor: 타입 지정, 동작 수정

- Input의 프로퍼티에 onFocus 추가
- non-disc Input에 Focus할 때 가상 키보드가 표시되지 않도록 수정
- Card의 bankName에 타입 지정

hyo-choi committed:
fix: non-disc Input에서 모달창이 사라지지 않는 문제 해결

hyo-choi committed:
refactor: state의 key를 constant로 대체

hyo-choi committed:
refactor: BankSelector 렌더링 방식 변경

hyo-choi committed:
chore: Storybook 그룹화, 정렬 및 누락분 추가

hyo-choi committed:
fix: BankItem의 onClick이 원모양에만 등록되도록 변경

cos18 commented on src/components/atoms/Input/Input.tsx:
```diff
@@ -0,0 +1,92 @@
+    className?.includes('card-register__input--card-owner')
+      ? '' : 'center'
+  )};
+  width: ${({ className }) => {
+    if (className?.includes('card-register__input--card-number')) {
+      return '4em';
+    } if (className?.includes('card-register__input--card-password')) {
+      return '1em';
+    } if (className?.includes('card-register__input--card-cvc')) {
+      return '3em';
+    } if (className?.includes('card-register__input--card-expiration')) {
+      return '5em';
+    } if (className?.includes('card-register__input--card-owner')) {
+      return '100%';
+    }
```
className으로 구분하는 것 보단 type이라는 프롭을 따로 만들어 구분하는게 좋아보입니다

hyo-choi replied:
리액트에 익숙하지 않아서 className을 사용하는 것이 보편적인 방식인 줄 알고 있었네요...! 중요한 피드백 감사합니다!

cos18 commented on src/components/atoms/Span/Span.tsx:
```diff
@@ -0,0 +1,36 @@
+
+const StyledSpan = styled.span`
+  display: inline-block;
+  font-size: ${({ className }) => (className?.includes('card-preview__span') ? 'small' : '')};
+  width: ${({ className }) => (
+    (className?.includes('card-preview__span--expiration')
+      || className?.includes('card-preview__span--owner-name'))
+      ? '50%' : '')};
+  text-align: ${({ className }) => (className?.includes('card-preview__span--expiration') ? 'right' : '')};
+  margin-left: ${({ className }) => (className?.includes('card-register__span--card-number') ? '0.5em' : '')};
+  margin-right: ${({ className }) => (className?.includes('card-register__span--card-number') ? '0.5em' : '')};
+`;
```
여기도 input 컴포넌트와 마찬가지로 type 프롭이 따로 있었으면 좋겠습니다~

cos18 commented on src/components/molecules/BankItem/BankItem.tsx:
```diff
@@ -0,0 +1,49 @@
+/* eslint-disable import/extensions */
+/* eslint-disable import/no-unresolved */
+/* eslint-disable arrow-body-style */
+// eslint-disable-next-line no-use-before-define
```
이런거는 eslint 설정파일에 rule로 작성해두면 좋을거같아요

cos18 commented on src/components/molecules/Modal/Modal.tsx:
```diff
@@ -0,0 +1,84 @@
+}: ModalProps) => {
+  return (
+    <>
+      <ModalBackground className={classnames(className, display ? '' : 'hidden')} onClick={onBackgroundClick} />
```
보통 모달의 백그라운드를 클릭하면 모달이 닫히는 암묵적인 룰이 있다보니 이를 기본기능으로 넣어두면 좋을것같아요

cos18 commented on src/components/molecules/Modal/Modal.tsx:
```diff
@@ -0,0 +1,84 @@
+  }
+`;
+
+const StyledModal = styled.div`
```
![Screen Shot 2021-07-11 at 2 09 32 AM](https://user-images.githubusercontent.com/33975709/125171063-26d62800-e1ed-11eb-80c5-7a23bb3d4da5.png)

모달 div 안에 패딩 넣었으면 좋겠어요...! 약간 삐죽 나온거 불편하네요 ㅋㅋㅋ

cos18 commented on src/components/organisms/CardRegisterForm/BankSelector/BankSelector.tsx:
```diff
@@ -0,0 +1,45 @@
+
+  return (
+    <>
+      <StyledDiv className={className}>
+        {bankItems.slice(0, 4)}
+      </StyledDiv>
+      <StyledDiv className={className}>
+        {bankItems.slice(4, 8)}
+      </StyledDiv>
```
이렇게 4개씩 나누는 작업을 slice를 사용해 처리하신건 좋은 것 같아요!

cos18 commented on src/components/organisms/CardRegisterForm/RegisterNumberInputs/RegisterNumberInputs.tsx:
```diff
@@ -0,0 +1,95 @@
+  classNames, label, values, refs, onChange, onFocus,
+}: CardRegisterNumberInputsProps) => {
+  return (
+    <StyledDiv>
```
![Screen Shot 2021-07-11 at 2 15 56 AM](https://user-images.githubusercontent.com/33975709/125171226-fc389f00-e1ed-11eb-9f24-6f8e61062f7b.png)

카드 번호의 input width가 부모 width의 100프로로 되어있어 사용자가 빨간 박스 밖을 눌러 입력하려 시도할수 있을 것 같습니다. 사이즈 자체를 빨간색 네모 박스만큼 줄이는게 좋지 않을까요?

cos18 commented on src/components/templates/CardRegisterForm/CardRegisterForm.tsx:
```diff
@@ -0,0 +1,324 @@
+  const [isModalDisplayed, setIsModalDisplayed] = useState(false);
+  const [modalContent, setModalContent] = useState<modalContentType>('BankSelector');
+  const [values, setValues] = useState<valueObjType>({
+    [FORM_KEY.CARD_BANK]: '',
+    [FORM_KEY.CARD_NUMBER]: '',
+    [FORM_KEY.CARD_NUMBER_FIRST]: '',
+    [FORM_KEY.CARD_NUMBER_SECOND]: '',
+    [FORM_KEY.CARD_NUMBER_THIRD]: '',
+    [FORM_KEY.CARD_NUMBER_FOURTH]: '',
+    [FORM_KEY.CARD_EXPIRATION]: '',
+    [FORM_KEY.CARD_CVC]: '',
+    [FORM_KEY.CARD_OWNER]: '',
+    [FORM_KEY.CARD_PASSWORD_FIRST]: '',
+    [FORM_KEY.CARD_PASSWORD_SECOND]: '',
```
굳이 이런 식으로 만들기보단 Card의 Interface나 class를 만들어서 그것의 생성자를 활용하는게 좋아보입니다. 당연히 useState<{[key: string]: string}>의 처리는 any를 사용하는 것과 거의 유사하기 때문에 타이핑의 의미가 없어집니다...

hyo-choi replied:
이 부분이 영 마음에 안 들었는데 어떻게 고쳐야 할지 감이 안 와서 그냥 뒀더니 짚어주셔서 정곡 찔린 기분이네요...! ``{[key:string]: string]}``은 확실히 별 의미 없는 타이핑이었던 것 같아요. 😅
그런데 Card의 Interface나 class를 만들어서 활용하라는 조언이 어떤 느낌인지 감이 잘 안 와서... 혹시 간단하게나마 예시를 좀 들어주실 수 있을까요?
(이건 사족이지만 매번 꼼꼼한 코드리뷰 정말!!!! 감사합니다! 아주 큰 도움이 되고 있어요... 😇)

cos18 reviewed (commented):
전체적으로 리뷰하는 내용 중 하나가 바로 className입니다. 리엑트를 사용하는 이유가 className을 최소한으로 사용하여 특정 컴포넌트를 지정할 수 있기 때문인데, 현재 코드는 소위 className이 덕지덕지 붙인 코드가 되어버렸습니다. 최대한 className 사용을 지양하고 코드를 짜 보는게 어떨까요?

hyo-choi committed:
chore: eslint rule 추가 및 disable 주석 일부 삭제

hyo-choi committed:
chore: Storybook 스토리 파일을 js에서 tsx로 변경

hyo-choi committed:
refactor: Modal component 수정

- content에 padding 추가
- className props 삭제

hyo-choi committed:
refactor: className 기반 스타일링을 prop 기반으로 변경

hyo-choi committed:
fix: 이전 커밋에서 누락된 변경사항 추가

hyo-choi committed:
feat: 가상 키보드 구현

Squashed commit of the following:

commit 34ba181682a3168587321cd99e6884854db5bac1
Author: hyo-choi <hyochoi31@gmail.com>
Date:   Mon Jul 12 00:22:42 2021 +0900

    feat: 가상 키보드 구현

commit 759e254e68c2f8e17aa51990d368648d05416450
Author: hyo-choi <hyochoi31@gmail.com>
Date:   Sun Jul 11 22:34:49 2021 +0900

    merge: main branch

commit 270dff704d4e1e803abdb1139aacff6169fae336
Author: hyo-choi <hyochoi31@gmail.com>
Date:   Sun Jul 11 17:02:26 2021 +0900

    feat: 가상 키보드 작업 중

hyo-choi committed:
chore: 일부 스타일 변경, eslint 주석 삭제
